controllers/cluster/cluster_controller.go

```diff
@@ -90,87 +88,27 @@ func (r *ClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_
 
 	log = log.WithValues("Cluster", cluster.Namespace+"/"+cluster.Name)
 
-	if _, exist := cluster.Labels[akoov1alpha1.AviClusterLabel]; !exist {
+	adcName, exist := cluster.Labels[akoov1alpha1.AviClusterLabel]
+	if !exist {
 		log.Info("Cluster doesn't have AVI enabled, skip Cluster reconciling")
 		return res, nil
 	}
 
 	log.Info("Cluster has AVI enabled, start Cluster reconciling")
 	// Getting all akodeploymentconfigs
-	var akoDeploymentConfigs akoov1alpha1.AKODeploymentConfigList
-	if err := r.Client.List(ctx, &akoDeploymentConfigs); err != nil {
-		return res, err
-	}
-
-	// Matches current cluster with all the akoDeploymentConfigs
-	clusterLabels := cluster.GetLabels()
-	matchedAkoDeploymentConfigs := make([]akoov1alpha1.AKODeploymentConfig, 0)
-	for _, akoDeploymentConfig := range akoDeploymentConfigs.Items {
-		if selector, err := metav1.LabelSelectorAsSelector(&akoDeploymentConfig.Spec.ClusterSelector); err != nil {
-			log.Error(err, "Failed to convert label sector to selector when matching ", "cluster", cluster.Name, " with ", akoDeploymentConfig.Name)
-		} else if selector.Matches(labels.Set(clusterLabels)) {
-			log.Info("Cluster ", cluster.Name, " is selected by", "Akodeploymentconfig", (akoDeploymentConfig.Namespace + "/" + akoDeploymentConfig.Name))
-			matchedAkoDeploymentConfigs = append(matchedAkoDeploymentConfigs, akoDeploymentConfig)
-		}
-	}
-
-	// When the cluster is not selected or only selected by the default ADC, a.k.a. install-ako-for-all
-	// We drop its skip-default-adc label, allowing cluster to use default ADC's config.
-	// It happens when a cluster is no longer selected by a customized ADC
-	if len(matchedAkoDeploymentConfigs) == 0 ||
-		(len(matchedAkoDeploymentConfigs) == 1 && matchedAkoDeploymentConfigs[0].Name == akoov1alpha1.WorkloadClusterAkoDeploymentConfig) {
-		delete(cluster.Labels, akoov1alpha1.AviClusterSelectedLabel)
-	}
+	var akoDeploymentConfig akoov1alpha1.AKODeploymentConfig
 
-	// If the cluster is selected by some AKODeploymentConfig objects, skip removing the finalizer
-	if len(matchedAkoDeploymentConfigs) > 0 {
+	if err := r.Client.Get(ctx, client.ObjectKey{Name: adcName}, &akoDeploymentConfig); err == nil {
 		return res, nil
-	}
-
-	// Removing finalizer if current cluster can't be selected by any akoDeploymentConfig
-	log.Info("Removing finalizer", "finalizer", akoov1alpha1.ClusterFinalizer)
-	ctrlutil.RemoveFinalizer(cluster, akoov1alpha1.ClusterFinalizer)
-
-	// Removing add on secret and its associated resources for a AKO
-	if _, err := r.deleteAddonSecret(ctx, log, cluster); err != nil {
-		log.Error(err, "Failed to remove secret", "secret", cluster.Name)
-		return res, err
-	}
-
-	// Removing avi label after deleting all the resources
-	delete(cluster.Labels, akoov1alpha1.AviClusterLabel)
+	} else if apierrors.IsNotFound(err) {
+		// Removing finalizer if current cluster can't be selected by any akoDeploymentConfig
```

Comment:
Is it ok to remove finalizer here? Should we match default adc instead?

Reply:
The `networking.tkg.tanzu.vmware.com/avi` gets updated in the mapper function, so if it does not exist, then it should not be selected by any adc object, it is safer to remove the finalizer.
Only the default adc with empty selector can select all the clusters, if default adc has some label selector, it should only select the cluster matching it's cluster selector.
We provide this flexibility because there is a scenario that users don't want to deploy ako in some specific workload cluster.